workflows/models.py

```diff
@@ -1,3 +1,170 @@
+from django.db import models
+from django.db.models.signals import pre_save
+from django.dispatch import receiver
+from django.utils.timezone import now
+from pydantic import BaseModel, ConfigDict, Field, field_validator
+
+from emgapiv2.model_utils import JSONFieldWithSchema
+
+from .prefect_utils.slurm_policies import _SlurmResubmitPolicy
+from .prefect_utils.slurm_status import SlurmStatus
 from .signals import ready
 
+
+class OrchestratedClusterJobManager(models.Manager):
+    def filter_similar_to_by_policy(
+        self,
+        policy: _SlurmResubmitPolicy,
+        job: "OrchestratedClusterJob.SlurmJobSubmitDescription",
+        input_file_hashes: List["OrchestratedClusterJob.JobInputFile"] = None,
+    ):
+        """
+        Filter for jobs with a similar job description, where similarity is determined
+        by the policy's list of field keys in `given_previous_job_matches`.
+        :param policy: A SlurmResubmitPolicy which defines the fields for matching.
+        :param job: A SlurmJobDescription to search for matches of.
+        :param input_file_hashes: A list of InputFileHash objects to match (if required by the policy).
+        :return: Filtered queryset.
+        """
+        filters = {
+            "created_at__lte": now() + policy.given_previous_job_submitted_before,
+            "created_at__gte": now() + policy.given_previous_job_submitted_after,
+        }
+        filters.update(
+            **{
+                f"job_submit_description__{field}": getattr(job, field)
+                for field in policy.given_previous_job_matches
+            }
+        )
+        if policy.considering_input_file_changes and input_file_hashes:
+            filters["input_files_hashes"] = [
+                ifh.model_dump() for ifh in input_file_hashes
+            ]
+
+        return self.get_queryset().filter(**filters).order_by("-created_at")
+
+    def get_previous_job(
+        self,
+        policy: _SlurmResubmitPolicy,
+        job: "OrchestratedClusterJob.SlurmJobSubmitDescription",
+        input_file_hashes: List["OrchestratedClusterJob.JobInputFile"],
+    ):
+        similar_past_jobs = self.filter_similar_to_by_policy(
+            policy, job, input_file_hashes
+        )
+
+        if similar_past_jobs.exists():
+            return similar_past_jobs.first()
+
+        return None
+
+
+class OrchestratedClusterJob(models.Model):
+    objects = OrchestratedClusterJobManager()
+
+    class SlurmJobSubmitDescription(BaseModel):
+        """
+        Pydantic schema for storing a slurm JobSubmitDescription object as json.
+        """
+
+        model_config = ConfigDict(extra="allow")
+
+        name: str = Field(..., description="Name of the job")
+        script: str = Field(..., description="Content of the job script/command")
+        memory_per_node: Optional[Union[str, int]] = Field(
+            None, description="Memory required for the job"
+        )
+        time_limit: Optional[str] = Field(
```

Comment:
Should this one be datetime?

Reply:
The idea here was to model the params that `pyslurm` takes (so same format as slurm CLI... and *hopefully* same as Slurm REST API – I need to look into that at some point). So `time_limit` is a string of format like `4-12:00:00` for 4.5 days. I thought about making a custom type for this based on `timedelta`, but decided it was slightly unneeded polish for now.

Reply:
agreed